tests/acceptance/30_custom_promise_types/11_handle.cf

```diff
@@ -0,0 +1,74 @@
+######################################################
+#
+#  Basic test of custom promise using the 'handle' attribute
+#
+#####################################################
+body common control
+{
+        inputs => { "../default.cf.sub" };
+        bundlesequence  => { default("$(this.promise_filename)") };
+        version => "1.0";
+}
+
+#######################################################
+
+bundle common python_check
+{
+  classes:
+    "python_version_compatible_with_cfengine_library"
+      expression => returnszero("/usr/bin/python3 -c 'import sys; assert sys.version_info >= (3,6)'", "useshell");
+@if minimum_version(3.17)
+    "cfengine_supports_custom_promises" expression => "any";
+@endif
+}
+
+bundle agent init
+{
+  meta:
+      "test_skip_unsupported" string => "!python_version_compatible_with_cfengine_library|!cfengine_supports_custom_promises";
+
+  files:
+      "$(this.promise_dirname)/cfengine.py"
+      copy_from => local_cp("$(this.promise_dirname)/../../../modules/promises/cfengine.py");
+}
+
+#######################################################
+
+@if minimum_version(3.17)
+promise agent multiline_insert
+{
+        interpreter => "/usr/bin/python3";
+        path => "$(this.promise_dirname)/multiline_insert.py";
+}
+
+bundle agent test
+{
+  meta:
+      "description" -> { "CFE-3439" }
+        string => "Test a custom promise using the 'handle' attribute";
+
+  multiline_insert:
+      "$(G.testfile)"
+        lines => { "handle is:", "$(this.handle)" },
+        handle => "example_handle";
```

Comment:
Every time I've seen handle used, it's for another promise to use in `depends_on`, so I would do that here or add another test. You could say that then you are testing depends_on as well, but I would say without it you are not testing the most important part of handle.

```suggestion
  multiline_insert:
      "$(G.testfile)"
        lines => { "handle is:", "$(this.handle)" },
        handle => "example_handle_1";
  classes:
    "found_example_handle_1" # Should be defined
      depends_on => "example_handle_1";
    "found_example_handle_2" # Shouldn't be defined
      depends_on => "example_handle_2";
```

Reply:
> Every time I've seen handle used, it's for another promise to use in `depends_on`, so I would do that here or add another test

It's also being used to trick the promise locking, but that's not implemented for custom promises at all.